mutex_yield.c

```diff
@@ -0,0 +1,73 @@
+#include <libkern.h>
+#include <mutex_yield.h>
+#include <sched.h>
+#include <critical_section.h>
+#include <clock.h>
+
+void mtx_yield_init(mtx_yield_t *mtx) {
+  assert(mtx);
+
+  *mtx = 0;
+}
+void mtx_yield_lock(mtx_yield_t *mtx) {
+  assert(mtx);
+
+  while (true) {
+    while (*mtx == 1)
+      sched_yield(true);
+
+    cs_enter();
+    if (*mtx == 0) {
+      *mtx = 1;
+      cs_leave();
+      return;
+    }
+    cs_leave();
+  }
+}
+void mtx_yield_unlock(mtx_yield_t *mtx) {
+  assert(mtx);
+
+  cs_enter();
+  *mtx = 0;
```

Comment:
It doesn't look like it is necessary to enter critical section here. Such simple, aligned, writes are usually guaranteed to be atomic. All that is needed is a compiler barrier so that memory accesses are not reordered after releasing the mutex.

Also, it is quite sad that such small functions cannot be inlined.